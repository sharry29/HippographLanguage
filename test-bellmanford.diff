1,3d0
< negative edge weight cycle
< negative edge weight cycle
< negative edge weight cycle
