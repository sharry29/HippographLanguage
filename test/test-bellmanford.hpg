int main() {
    graph<string:int, int> g = ["S":500 -(10)> "A":500 -(2)> "C":500 -(-2)> "B" -(1)> "A"; "S" -(8)>"E":500 -(1)> "D":500 -(-1)>"C"; "D"-(-4)>"A"];

    graph<string:int, int> shortest_paths = [];
    bool valid = true;

    for_node(n : g) {
        node<string:int> n1 = n.get_name():999999;
        shortest_paths.set_node(n1);
    }

    for_node(n : shortest_paths) {
        for_edge(src, dst, w : g) {
            int src_data = src.get_data();
            int dst_data = dst.get_data();
            if (src_data + w < dst_data) {
                dst.set_data(src_data + w);
                shortest_paths.set_node(dst);
                shortest_paths.set_edge(src.get_name(); dst.get_name(); 0);
            }
        }
    }

    for_edge(src, dst, w : g) {
        int src_data = src.get_data();
        int dst_data = dst.get_data();
        if (src_data + w < dst_data) {
            print("negative edge weight cycle");
            valid = false;
        }
    }

    if (valid) {
        shortest_paths.print();
    }

    return 0;
}
