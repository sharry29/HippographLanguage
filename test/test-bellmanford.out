ORIGINAL GRAPH:
"S":500 -> ["A":500 (10), "E":500 (8)]
"A":500 -> ["C":500 (2)]
"C":500 -> ["B":500 (2)]
"B":500 -> ["A":500 (1)]
"E":500 -> ["D":500 (1)]
"D":500 -> ["C":500 (1), "A":500 (4)]
SHORTEST PATH:
"S":0 -> ["A":10 (10), "E":8 (8)]
"A":10 -> []
"C":10 -> ["B":12 (2)]
"B":12 -> []
"E":8 -> ["D":9 (1)]
"D":9 -> ["C":10 (1)]
